connectors/src/connectors/github/temporal/activities.ts

```diff
@@ -711,3 +728,261 @@ export function getDiscussionDocumentId(
+
+    const codeSyncStartedAt = new Date();
+    const rootInternalId = `github-code-${repoId}`;
+
+    const updatedDirectories: { [key: string]: boolean } = {};
+
+    for (const f of files) {
+      // Read file (files are 1MB at most).
+      const content = await fs.readFile(f.localFilePath);
+      const contentHash = blake3(content).toString("hex");
+
+      // Find file or create it with an empty contentHash.
+      let githubCodeFile = await GithubCodeFile.findOne({
+        where: {
+          connectorId: connector.id,
+          repoId,
+          documentId: f.documentId,
+        },
+      });
+
+      if (!githubCodeFile) {
+        githubCodeFile = await GithubCodeFile.create({
+          connectorId: connector.id,
+          repoId,
+          documentId: f.documentId,
+          parentInternalId: f.parentInternalId || rootInternalId,
+          fileName: f.fileName,
+          sourceUrl: f.sourceUrl,
+          contentHash: "",
+          createdAt: codeSyncStartedAt,
+          updatedAt: codeSyncStartedAt,
+          lastSeenAt: codeSyncStartedAt,
+        });
+      }
+
+      // We update if the file name or source url or content has changed.
+      const needsUpdate =
+        f.fileName !== githubCodeFile.fileName ||
+        f.parentInternalId !== githubCodeFile.parentInternalId ||
+        f.sourceUrl !== githubCodeFile.sourceUrl ||
+        contentHash === githubCodeFile.contentHash;
+
+      if (needsUpdate) {
+        // Record the parent directories to update their updatedAt.
+        for (const parentInternalId of f.parents) {
+          updatedDirectories[parentInternalId] = true;
+        }
+
+        const tags = [
+          `title:${f.fileName}`,
+          `lasUpdatedAt:${codeSyncStartedAt.getTime()}`,
+        ];
+
+        // Time to upload the file to the data source.
+        await upsertToDatasource({
+          dataSourceConfig,
+          documentId: f.documentId,
+          documentContent: formatCodeContentForUpsert(f.sourceUrl, content),
+          documentUrl: f.sourceUrl,
+          timestampMs: codeSyncStartedAt.getTime(),
+          tags,
+          parents: [...f.parents, rootInternalId, repoId],
+          retries: 3,
+          delayBetweenRetriesMs: 1000,
+          loggerArgs: { ...loggerArgs, provider: "github" },
+          upsertContext: {
+            sync_type: isBatchSync ? "batch" : "incremental",
+          },
+        });
+
+        // Finally update the file.
+        githubCodeFile.fileName = f.fileName;
+        githubCodeFile.parentInternalId = f.parentInternalId || rootInternalId;
+        githubCodeFile.sourceUrl = f.sourceUrl;
+        githubCodeFile.contentHash = contentHash;
+        githubCodeFile.updatedAt = codeSyncStartedAt;
+      }
+
+      // Finally we update the lastSeenAt for all files we've seen, and save.
+      githubCodeFile.lastSeenAt = codeSyncStartedAt;
+      await githubCodeFile.save();
+    }
+
+    for (const d of directories) {
```

Comment:
Any specific reason we don't want to crank up the concurrency a little bit here ?

Reply:
We can but that will run very quickly won't it?

Reply:
What would be the value of batching?